File: python/a_star.py

import math
def heuristic(current_node, goal_node):
    dx = current_node.x - goal_node.x
    dy = current_node.y - goal_node.y
    dz = (current_node.z - goal_node.z) * 10 # Multiplier compensates for floor height
    return math.sqrt(dx**2 + dy**2 + dz**2)

import heapq

def a_star(graph, start_node, goal_node):
    """
    Executes the A* pathfinding algorithm.
    Returns a list of Node objects representing the shortest path, 
    or None if no path exists.
    """
    # Priority queue stores tuples: (f_score, unique_id, node)
    # The unique_id (using id()) prevents Python from crashing if two nodes 
    # have the exact same f_score, as it won't try to compare the Node objects directly.
    open_set = []
    heapq.heappush(open_set, (0, id(start_node), start_node))
    
    # Dictionary to reconstruct the winning path later
    came_from = {}
    
    # g_score: The exact, known cost from the start node to the current node.
    g_score = {start_node: 0}
    
    # f_score: g_score + estimated cost to the goal (heuristic).
    f_score = {start_node: heuristic(start_node, goal_node)}
    
    # A fast-lookup set to check if a node is currently in the priority queue
    open_set_hash = {start_node}
    
    while open_set:
        # 1. Get the node with the absolute lowest f_score
        current_tuple = heapq.heappop(open_set)
        current = current_tuple[2]
        open_set_hash.remove(current)
        
        # 2. Check if we found the destination
        if current == goal_node:
            # We arrived! Now walk backwards to build the final path list
            path = []
            while current in came_from:
                path.append(current)
                current = came_from[current]
            path.append(start_node)
            return path[::-1] # Reverse the list so it goes from Start -> Goal
            
        # 3. Explore all valid connected edges (neighbors)
        for neighbor_id, move_cost in current.edges.items():
            
            # Fetch the actual Node object on the fly!
            neighbor = graph.nodes[neighbor_id]

            # Calculate what the g_score WOULD be if we took this path
            tentative_g_score = g_score[current] + move_cost
            
            # If this is the fastest way we've ever reached this neighbor...
            if tentative_g_score < g_score.get(neighbor, float('inf')):
                
                # ...save this route!
                came_from[neighbor] = current
                g_score[neighbor] = tentative_g_score
                f_score[neighbor] = tentative_g_score + heuristic(neighbor, goal_node)
                
                # If we haven't already queued this neighbor for exploration, add it
                if neighbor not in open_set_hash:
                    heapq.heappush(open_set, (f_score[neighbor], id(neighbor), neighbor))
                    open_set_hash.add(neighbor)
                    
    # If the while loop completely empties the queue and we never hit the goal, 
    # it means the destination is physically trapped behind walls.
    return None

File: typescript/a_star.ts

import { Node, Graph } from './graph.js';

/**
 * MinHeap implementation to replace Python's heapq.
 * Stores tuples of [fScore, tieBreaker, Node].
 * Compares by fScore first, then tieBreaker to avoid object comparison.
 */
type HeapEntry = [number, number, Node];

class MinHeap {
    private heap: HeapEntry[] = [];

    push(entry: HeapEntry): void {
        this.heap.push(entry);
        this._bubbleUp(this.heap.length - 1);
    }

    pop(): HeapEntry {
        const top = this.heap[0];
        const last = this.heap.pop()!;
        if (this.heap.length > 0) {
            this.heap[0] = last;
            this._bubbleDown(0);
        }
        return top;
    }

    get length(): number {
        return this.heap.length;
    }

    private _compare(a: HeapEntry, b: HeapEntry): number {
        if (a[0] !== b[0]) return a[0] - b[0];
        return a[1] - b[1];
    }

    private _bubbleUp(idx: number): void {
        while (idx > 0) {
            const parentIdx = Math.floor((idx - 1) / 2);
            if (this._compare(this.heap[idx], this.heap[parentIdx]) < 0) {
                [this.heap[idx], this.heap[parentIdx]] = [this.heap[parentIdx], this.heap[idx]];
                idx = parentIdx;
            } else {
                break;
            }
        }
    }

    private _bubbleDown(idx: number): void {
        const length = this.heap.length;
        while (true) {
            let smallest = idx;
            const left = 2 * idx + 1;
            const right = 2 * idx + 2;

            if (left < length && this._compare(this.heap[left], this.heap[smallest]) < 0) {
                smallest = left;
            }
            if (right < length && this._compare(this.heap[right], this.heap[smallest]) < 0) {
                smallest = right;
            }
            if (smallest !== idx) {
                [this.heap[idx], this.heap[smallest]] = [this.heap[smallest], this.heap[idx]];
                idx = smallest;
            } else {
                break;
            }
        }
    }
}

export function heuristic(currentNode: Node, goalNode: Node): number {
    const dx = currentNode.x - goalNode.x;
    const dy = currentNode.y - goalNode.y;
    const dz = (currentNode.z - goalNode.z) * 10; // Multiplier compensates for floor height
    return Math.sqrt(dx ** 2 + dy ** 2 + dz ** 2);
}

let heapCounter = 0;

export function aStar(graph: Graph, startNode: Node, goalNode: Node): Node[] | null {
    /**
     * Executes the A* pathfinding algorithm.
     * Returns a list of Node objects representing the shortest path,
     * or null if no path exists.
     */
    // Priority queue stores tuples: (f_score, unique_id, node)
    // The unique_id (using heapCounter) prevents crashes if two nodes
    // have the exact same f_score, as it won't try to compare the Node objects directly.
    heapCounter = 0;
    const openSet = new MinHeap();
    heapqPush(openSet, 0, startNode);

    // Dictionary to reconstruct the winning path later
    const cameFrom = new Map<Node, Node>();

    // g_score: The exact, known cost from the start node to the current node.
    const gScore = new Map<Node, number>();
    gScore.set(startNode, 0);

    // f_score: g_score + estimated cost to the goal (heuristic).
    const fScore = new Map<Node, number>();
    fScore.set(startNode, heuristic(startNode, goalNode));

    // A fast-lookup set to check if a node is currently in the priority queue
    const openSetHash = new Set<Node>();
    openSetHash.add(startNode);

    while (openSet.length > 0) {
        // 1. Get the node with the absolute lowest f_score
        const currentTuple = openSet.pop();
        let current = currentTuple[2];
        openSetHash.delete(current);

        // 2. Check if we found the destination
        if (current === goalNode) {
            // We arrived! Now walk backwards to build the final path list
            const path: Node[] = [];
            while (cameFrom.has(current)) {
                path.push(current);
                current = cameFrom.get(current)!;
            }
            path.push(startNode);
            return path.reverse(); // Reverse the list so it goes from Start -> Goal
        }

        // 3. Explore all valid connected edges (neighbors)
        for (const [neighborId, moveCost] of Object.entries(current.edges)) {

            // Fetch the actual Node object on the fly!
            const neighbor = graph.nodes[neighborId];

            // Calculate what the g_score WOULD be if we took this path
            const tentativeGScore = (gScore.get(current) ?? Infinity) + moveCost;

            // If this is the fastest way we've ever reached this neighbor...
            if (tentativeGScore < (gScore.get(neighbor) ?? Infinity)) {

                // ...save this route!
                cameFrom.set(neighbor, current);
                gScore.set(neighbor, tentativeGScore);
                fScore.set(neighbor, tentativeGScore + heuristic(neighbor, goalNode));

                // If we haven't already queued this neighbor for exploration, add it
                if (!openSetHash.has(neighbor)) {
                    heapqPush(openSet, fScore.get(neighbor)!, neighbor);
                    openSetHash.add(neighbor);
                }
            }
        }
    }

    // If the while loop completely empties the queue and we never hit the goal,
    // it means the destination is physically trapped behind walls.
    return null;
}

/** Helper to mirror heapq.heappush with auto-incrementing tiebreaker */
function heapqPush(heap: MinHeap, fScore: number, node: Node): void {
    heap.push([fScore, heapCounter++, node]);
}
